### PR TITLE
Support SM4

### PR DESCRIPTION
**Pull request by JerryDevis**
Support SM4 algorithm

### CODE REVIEW

**Comment by stefanberger** [comment]
I wonder whether the TPM 2 code is in general prepared for the type of expansion of plaintext versus ciphertext. In the example 32 bytes input with SM4 become 48 bytes output. My concern would be that this could lead to buffer overruns or failure mode if large input data are given.

**Comment by JerryDevis** [comment]
> This fixes the test issue it seems:
> 
> ```
> diff --git a/src/tpm2/SymmetricTest.h b/src/tpm2/SymmetricTest.h
> index 058f50a..8083191 100644
> --- a/src/tpm2/SymmetricTest.h
> +++ b/src/tpm2/SymmetricTest.h
> @@ -96,7 +96,7 @@ const SYMMETRIC_TEST_VECTOR   c_symTestValues[NUM_SYMS + 1] = {
>  #if ALG_SM4 && SM4_128 // libtpms activated
>      {TPM_ALG_SM4, 128, key_SM4128, 16, sizeof(dataIn_SM4128), dataIn_SM4128,
>       {dataOut_SM4128_CTR, dataOut_SM4128_OFB, dataOut_SM4128_CBC,
> -      dataOut_SM4128_CFB, dataOut_AES128_ECB}},
> +      dataOut_SM4128_CFB, dataOut_SM4128_ECB}},
>  #endif
>  // libtpms added begin
>  #if ALG_TDES && TDES_128
> ```
Ok，I have fixed it.

File: src/tpm2/Marshal_fp.h [hunk 1]
```diff
@@ -238,8 +238,12 @@ extern "C" {
     TPM2B_ATTEST_Marshal(TPM2B_ATTEST *source, BYTE **buffer, INT32 *size);
     UINT16
     TPMI_AES_KEY_BITS_Marshal(TPMI_AES_KEY_BITS *source, BYTE **buffer, INT32 *size);
-    UINT16			// libtpms added
-    TPMI_TDES_KEY_BITS_Marshal(TPMI_TDES_KEY_BITS *source, BYTE **buffer, INT32 *size);
+#if ALG_SM4 			// libtpms begin
```
Comment (stefanberger) [inline src/tpm2/Marshal_fp.h:241]: libtpms added begin

File: src/tpm2/crypto/CryptSym.h [hunk 2]
```diff
@@ -123,6 +123,7 @@ typedef union tpmCryptKeySchedule_t {
     encrypt(SWIZZLE(keySchedule, in, out))
 #   define DECRYPT(keySchedule, in, out)	\
     decrypt(SWIZZLE(keySchedule, in, out))
+#define FINAL(keySchedule) final((void *)(keySchedule))
```
Comment (stefanberger) [inline src/tpm2/crypto/CryptSym.h:126]: libtpms added

File: src/tpm2/crypto/CryptSym.h [hunk 3]
```diff
@@ -132,11 +133,13 @@ typedef union tpmCryptKeySchedule_t {
     case TPM_ALG_##ALG:							\
     TpmCryptSetEncryptKey##ALG(key, keySizeInBits, &keySchedule.alg);	\
     encrypt = (TpmCryptSetSymKeyCall_t)TpmCryptEncrypt##ALG;		\
+    final = (TpmCryptSymFinal_t)TpmCryptFinal##ALG;          \
```
Comment (stefanberger) [inline src/tpm2/crypto/CryptSym.h:136]: libtpms added

File: src/tpm2/crypto/CryptSym.h [hunk 4]
```diff
@@ -132,11 +133,13 @@ typedef union tpmCryptKeySchedule_t {
...
     break;
 #define DECRYPT_CASE(ALG, alg)						\
     case TPM_ALG_##ALG:							\
     TpmCryptSetDecryptKey##ALG(key, keySizeInBits, &keySchedule.alg);	\
     decrypt = (TpmCryptSetSymKeyCall_t)TpmCryptDecrypt##ALG;		\
+    final = (TpmCryptSymFinal_t)TpmCryptFinal##ALG;          \
```
Comment (stefanberger) [inline src/tpm2/crypto/CryptSym.h:142]: libtpms added

File: src/tpm2/crypto/openssl/CryptCmac.c [hunk 5]
```diff
@@ -119,7 +119,8 @@ CryptCmacData(
     BYTE                    *key = cmacState->symKey.t.buffer;
     UINT16                   keySizeInBits = cmacState->keySizeBits;
     tpmCryptKeySchedule_t    keySchedule;
-    TpmCryptSetSymKeyCall_t  encrypt;
+    TpmCryptSetSymKeyCall_t  encrypt;
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptCmac.c:122]: it looks like you added a whitespace to the end of this line...

File: src/tpm2/crypto/openssl/CryptCmac.c [hunk 6]
```diff
@@ -119,7 +119,8 @@ CryptCmacData(
     BYTE                    *key = cmacState->symKey.t.buffer;
     UINT16                   keySizeInBits = cmacState->keySizeBits;
     tpmCryptKeySchedule_t    keySchedule;
-    TpmCryptSetSymKeyCall_t  encrypt;
+    TpmCryptSetSymKeyCall_t  encrypt; 
+    TpmCryptSymFinal_t       final;
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptCmac.c:123]: comment: /* libtpms added */

File: src/tpm2/crypto/openssl/CryptCmac.c [hunk 7]
```diff
@@ -142,6 +143,8 @@ CryptCmacData(
 	            cmacState->iv.t.buffer[cmacState->bcount] ^= *buffer++;
 	        }
 	}
+    if (final)
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptCmac.c:146]: comment: /* libtpms added begin */   and then /* lbtpms added end */

File: src/tpm2/crypto/openssl/CryptCmac.c [hunk 8]
```diff
@@ -162,6 +165,7 @@ CryptCmacEnd(
     UINT16                   keySizeInBits = cState->keySizeBits;
     tpmCryptKeySchedule_t    keySchedule;
     TpmCryptSetSymKeyCall_t  encrypt;
+    TpmCryptSymFinal_t       final;
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptCmac.c:168]: /* libtpms added */

File: src/tpm2/crypto/openssl/CryptCmac.c [hunk 9]
```diff
@@ -204,6 +208,8 @@ CryptCmacEnd(
     i = (UINT16)MIN(cState->iv.t.size, outSize);
     MemoryCopy(outBuffer, cState->iv.t.buffer, i);
 
+    if (final)
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptCmac.c:211]: comment: /* libtpms added begin / and then / lbtpms added end */

File: src/tpm2/crypto/openssl/CryptSym.c [hunk 10]
```diff
@@ -175,6 +175,7 @@ CryptSymmetricEncrypt(
     tpmCryptKeySchedule_t        keySchedule;
     INT16                blockSize;
     TpmCryptSetSymKeyCall_t        encrypt;
+    TpmCryptSymFinal_t       final;
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptSym.c:178]: comment: /* libtpms added */

File: src/tpm2/crypto/openssl/CryptSym.c [hunk 11]
```diff
@@ -294,8 +295,12 @@ CryptSymmetricEncrypt(
 	    break;
 #endif
 	  default:
+	  	if (final)
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptSym.c:298]: comment ...

File: src/tpm2/crypto/openssl/CryptSym.c [hunk 12]
```diff
@@ -326,6 +331,7 @@ CryptSymmetricDecrypt(
     BYTE                *iv;
     TpmCryptSetSymKeyCall_t        encrypt;
     TpmCryptSetSymKeyCall_t        decrypt;
+	TpmCryptSymFinal_t       final;
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptSym.c:334]: comment ...

File: src/tpm2/crypto/openssl/CryptSym.c [hunk 13]
```diff
@@ -466,8 +472,12 @@ CryptSymmetricDecrypt(
 	    break;
 #endif
 	  default:
+	    if (final)
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/CryptSym.c:475]: comment ...

File: src/tpm2/crypto/openssl/TpmToOsslSym.h [hunk 14]
```diff
@@ -138,6 +135,7 @@ typedef void(*TpmCryptSetSymKeyCall_t)(
 #define TpmCryptEncryptAES          AES_encrypt
 #define TpmCryptDecryptAES          AES_decrypt
 #define tpmKeyScheduleAES           AES_KEY
+#define TpmCryptFinalAES            NULL
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/TpmToOsslSym.h:138]: comment ...

File: src/tpm2/crypto/openssl/TpmToOsslSym.h [hunk 15]
```diff
@@ -165,14 +164,15 @@ typedef void(*TpmCryptSetSymKeyCall_t)(
 /* Macros to set up the encryption/decryption key schedules */
 
 #define TpmCryptSetEncryptKeySM4(key, keySizeInBits, schedule)	\
-    SM4_set_key((key), (tpmKeyScheduleSM4 *)(schedule))
+    SM4_set_encrypt_key((key), (tpmKeyScheduleSM4 *)(schedule))
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/TpmToOsslSym.h:167]: /* libtpms changed */

File: src/tpm2/crypto/openssl/TpmToOsslSym.h [hunk 16]
```diff
@@ -165,14 +164,15 @@ typedef void(*TpmCryptSetSymKeyCall_t)(
...
 #define TpmCryptSetDecryptKeySM4(key, keySizeInBits, schedule)	\
-    SM4_set_key((key), (tpmKeyScheduleSM4 *)(schedule))
+    SM4_set_decrypt_key((key), (tpmKeyScheduleSM4 *)(schedule))
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/TpmToOsslSym.h:169]: /* libtpms changed */

File: src/tpm2/crypto/openssl/TpmToOsslSym.h [hunk 17]
```diff
@@ -156,6 +154,7 @@ typedef void(*TpmCryptSetSymKeyCall_t)(
 #define TpmCryptEncryptTDES         TDES_encrypt
 #define TpmCryptDecryptTDES         TDES_decrypt
 #define tpmKeyScheduleTDES          DES_key_schedule
+#define TpmCryptFinalTDES           NULL
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/TpmToOsslSym.h:157]: comment ...

File: src/tpm2/crypto/openssl/TpmToOsslSym.h [hunk 18]
```diff
@@ -165,14 +164,15 @@ typedef void(*TpmCryptSetSymKeyCall_t)(
...
+    SM4_set_encrypt_key((key), (tpmKeyScheduleSM4 *)(schedule))
 #define TpmCryptSetDecryptKeySM4(key, keySizeInBits, schedule)	\
-    SM4_set_key((key), (tpmKeyScheduleSM4 *)(schedule))
+    SM4_set_decrypt_key((key), (tpmKeyScheduleSM4 *)(schedule))
 /* Macros to alias encryption calls to specific algorithms. This should be used sparingly. */
 
 #define TpmCryptEncryptSM4          SM4_encrypt
 #define TpmCryptDecryptSM4          SM4_decrypt
 #define tpmKeyScheduleSM4           SM4_KEY
+#define TpmCryptFinalSM4            SM4_final
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/TpmToOsslSym.h:175]: comment ...

File: src/tpm2/crypto/openssl/TpmToOsslSym.h [hunk 19]
```diff
@@ -187,6 +187,7 @@ typedef void(*TpmCryptSetSymKeyCall_t)(
 #define TpmCryptEncryptCAMELLIA          Camellia_encrypt
 #define TpmCryptDecryptCAMELLIA          Camellia_decrypt
 #define tpmKeyScheduleCAMELLIA           CAMELLIA_KEY
+#define TpmCryptFinalCAMELLIA            NULL
```
Comment (stefanberger) [inline src/tpm2/crypto/openssl/TpmToOsslSym.h:190]: comment ...

File: src/tpm2/AlgorithmTests.c [hunk 20]
```diff
@@ -845,7 +845,7 @@ TestAlgorithm(
 #if ALG_SM4
 		    // if SM4 is implemented, its test is like other block ciphers but there
 		    // aren't any test vectors for it yet
-		    //            case TPM_ALG_SM4:
+		  case TPM_ALG_SM4:
```
Comment (stefanberger) [inline src/tpm2/AlgorithmTests.c:848]: comment: /* libtpms changed */

**Review by stefanberger — CHANGES_REQUESTED** [review]
It's mostly about adding these comments to the code now and removing white spaces. I have the comments in many other places and they help me skip over this stuff when synchronizing with the reference code. Ideally they would all disappear at some point when the reference code has the SM 4 support as well...

**Comment by JerryDevis** [comment]
There are also some problems with implementing these algorithms ourselves.
Firstly, these algorithms, which we implemented ourselves, lack the certification of some security organizations.
Secondly, OpenSSL or other cryptographic library suites use some assembly language to accelerate the algorithms, If we do not consider these, then the performance of our implemented algorithm will be poor.
For the future, this is indeed a very troublesome problem.

**Comment by fly2x** [comment]
JerryDevis is on leave. We're from the same team. Once you finish the re-base, we'll resubmit.

**Pull request merged**